modules/delegates.js

```diff
@@ -720,23 +688,27 @@ Delegates.prototype.updateForgingStatus = function(
 /**
  * Gets delegate list based on input function by vote and changes order.
  *
- * @param {number} round
- * @param {function} source - Source function for get delegates
+ * @param {number} blockHeight
  * @param {function} cb - Callback function
  * @param {Object} tx - Database transaction/task object
  * @returns {setImmediateCallback} cb, err, truncated delegate list
  * @todo Add description for the params
  */
-Delegates.prototype.generateDelegateList = function(round, source, cb, tx) {
-	// Set default function for getting delegates
-	source = source || __private.getKeysSortByVote;
+Delegates.prototype.generateDelegateList = function(blockHeight, cb, tx) {
+	const round = slots.calcRound(blockHeight);
 
 	if (__private.delegatesListCache[round]) {
-		library.logger.debug('Using delegate list from the cache for round', round);
 		return setImmediate(cb, null, __private.delegatesListCache[round]);
 	}
 
-	return source((err, truncDelegateList) => {
+	const lastBlockHeight = modules.blocks.lastBlock.get().height;
+	// Set default function for getting delegates
+	const source =
+		blockHeight === lastBlockHeight && lastBlockHeight % ACTIVE_DELEGATES === 0
+			? 'getDelegatesFromPreviousRound'
+			: 'getKeysSortByVote';
```

Comment:
I would rename this method to `getDelegatesForCurrentRound`

Reply:
@4miners could you please comment on this ?